### DISCUSSION TITLE
🚀 Born v0.5.0 Released - LLM Support!

### DISCUSSION THREAD

**Discussion by kolkov** [born-ml/born#8] [Announcements]
# Born v0.5.0 - LLM Support

We're excited to announce **Born v0.5.0**, bringing everything you need to run Large Language Models in pure Go!

### 🔤 Tokenizer Package
Full tokenization support for modern LLMs:
- **BPE Tokenizer** - Byte Pair Encoding with regex pre-tokenization
- **Tiktoken** - OpenAI-compatible tokenizer (GPT-3.5/4)
- **HuggingFace Loader** - Load tokenizers directly from HF format
- **Chat Templates** - Jinja2-style template rendering for chat models

### 📦 Model Loader
Load pre-trained weights from popular formats:
- **SafeTensors** - Memory-efficient, safe tensor format
- **GGUF** - llama.cpp quantized model support
- **Automatic Architecture Detection** - LLaMA, Mistral, Phi, Qwen, and more

### 🎯 Text Generation
Production-ready inference:
- **Sampling Strategies** - Greedy, Top-K, Top-P (nucleus), Temperature
- **KV Cache Ready** - Prepared for efficient autoregressive generation
- **Streaming Support** - Token-by-token output

### 🧠 Neural Network Modules
Modern transformer components:
- **GLU Variants** - SwiGLU, GELU, GeGLU, ReGLU, Bilinear
- **Grouped Query Attention (GQA)** - Efficient attention for large models
- **SwiGLU FFN** - LLaMA-style feed-forward networks

### 🔧 Autodiff Improvements
Critical fixes for training:
- Where operation backward pass
- Multi-output operations (Chunk) support
- Unsqueeze/Squeeze tape recording

## Installation

```bash
go get github.com/born-ml/born@v0.5.0
```

## Stats
- 📁 48 files changed
- ➕ 9,391 lines of code
- ✅ All tests passing (Linux, macOS, Windows)

## What's Next
- v0.6.0: KV Cache implementation
- v0.7.0: GPU acceleration (Vulkan)
- v1.0.0: Production-ready release

---

**Full changelog:** https://github.com/born-ml/born/releases/tag/v0.5.0

Let us know what you think! 🎉

**Comment by kolkov** [comment]
## 🔧 v0.5.1 Hotfix Released

A critical bug was discovered and fixed:

**Problem**: `nn.Embedding` was not recording operations on the autodiff tape, causing zero gradients for embedding weights.

**Solution**: Embedding lookup now routes through `backend.Embedding()` which properly records on tape.

**Upgrade**:
```bash
go get github.com/born-ml/born@v0.5.1
```

If you're using embeddings (transformers, LLMs, etc.), please update to v0.5.1!

**Comment by kolkov** [comment]
## v0.5.2 Released!

**Critical Fixes:**
- `nn.CrossEntropyLoss` now records on autodiff tape (gradients flow properly)
- Public WebGPU API: `import "github.com/born-ml/born/backend/webgpu"`
- `webgpu.IsAvailable()` for graceful GPU fallback
- +96 tensor tests (core coverage: 93.2%)

https://github.com/born-ml/born/releases/tag/v0.5.2
